File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.py
from collections import deque

N, K = map(int, input().split())
MAX = 100001
dist = [-1] * MAX  # -1로 초기화하여 방문 여부를 체크합니다.

def bfs(start):
    queue = deque([start])
    dist[start] = 0  # 시작점의 거리 초기화

    while queue:
        x = queue.popleft()

        if x == K:
            return dist[x]

        # x * 2의 경우, 거리를 업데이트 하지 않으므로, 최단 거리로 가능한 경우 가장 먼저 방문할 수 있도록 처리합니다.
        if 0 <= x * 2 < MAX and dist[x * 2] == -1:
            dist[x * 2] = dist[x]  # 이동 비용이 0
            queue.append(x * 2)

        # x - 1의 경우, 이동 비용이 1
        if 0 <= x - 1 < MAX and dist[x - 1] == -1:
            dist[x - 1] = dist[x] + 1
            queue.append(x - 1)

        # x + 1의 경우, 이동 비용이 1
        if 0 <= x + 1 < MAX and dist[x + 1] == -1:
            dist[x + 1] = dist[x] + 1
            queue.append(x + 1)

print(bfs(N))

File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int N, K;
    static Integer[] dist;
    static int[] dx = {-1, 1};
    static Queue<Integer> q = new LinkedList<>();

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());

        dist = new Integer[100001];

        dist[N] = 0;
        q.add(N);
        bfs(q);

        System.out.print(dist[K]);
    }

    public static void bfs(Queue<Integer> q) {
        while(!q.isEmpty()) {
            int x = q.poll();

            if (x == K) {
                return;
            }

            if (x * 2 >= 0 && x * 2 < 100001 && dist[x * 2] == null) {
                dist[x * 2] = dist[x];
                q.add(x * 2);
            }

            for (int i = 0; i < 2; i++) {
                int nx = x + dx[i];

                if (nx >= 0 && nx < 100001 && dist[nx] == null) {
                    dist[nx] = dist[x] + 1;
                    q.add(nx);
                }
            }
        }
    }
}
